src/main/scala/is/hail/methods/PCRelate.scala

```diff
@@ -109,45 +109,52 @@ object PCRelate {
     val nSamples = vds.nSamples
     val variants = vds.variants.collect()
     val variantIdxBc = vds.sparkContext.broadcast(variants.index)
-    val rdd = vds.rdd.mapPartitions { part =>
-      part.map { case (v, (va, gs)) =>
+    val localRowType = vds.rowType
+    val rdd = vds.rdd2.mapPartitions { it =>
+      val view = HardCallView(localRowType)
+      val missingIndices = new ArrayBuilder[Int]()
+
+      it.map { rv =>
+        val v = Variant.fromRegionValue(rv.region,
+          localRowType.loadField(rv, 1))
+        view.setRegion(rv)
+
+        missingIndices.clear()
         var sum = 0
         var nNonMissing = 0
-        val missingIndices = new ArrayBuilder[Int]()
         val a = new Array[Double](nSamples)
-
         var i = 0
-        val it = gs.hardCallIterator
-        while (it.hasNext) {
-          val gt = it.next()
-          if (gt == -1) {
-            missingIndices += i
-          } else {
+        while (i < nSamples) {
+          view.setGenotype(i)
+          if (view.hasGT) {
+            val gt = view.getGT
             sum += gt
             a(i) = gt
             nNonMissing += 1
-          }
+          } else
+            missingIndices += i
           i += 1
         }
 
         val mean = sum.toDouble / nNonMissing
 
-        for (i <- missingIndices.result()) {
-          a(i) = mean
+        i = 0
+        while (i < missingIndices.length) {
+          a(missingIndices(i)) = mean
+          i += 1
         }
 
-        // FIXME: this should probably be a sparse vector
-        new IndexedRow(variantIdxBc.value(v), new DenseVector(a))
+        IndexedRow(variantIdxBc.value(v), new DenseVector(a))
```

Comment:
The `variantIdxBc.value(v)` here is the only use of both `variantIdxBc` and `v`. Could the `zipWithIndex` method on `RDD` be used to avoid the allocations? It looks to me like it would give the same result.

Reply:
You're totally right, good call.  I'm going to merge this now so it doesn't block killing Genotype, but I put this on my TODO list and will make a second PR.